bench/bench.ml

```diff
@@ -0,0 +1,484 @@
+  type result = {
+    time : float;
+    ops_per_sec : float;
+    mbs_per_sec : float;
+    read_amplification_calls : float;
+    read_amplification_size : float;
+    write_amplification_calls : float;
+    write_amplification_size : float;
+  }
+  [@@deriving yojson]
+
+  let run ~nb_entries f =
+    let _, time, stats = with_stats (fun () -> f ()) in
+    let nb_entriesf = float_of_int nb_entries in
+    let entry_sizef = float_of_int entry_size in
+    let read_amplification_size =
+      float_of_int stats.bytes_read /. (entry_sizef *. nb_entriesf)
+    in
+    let read_amplification_calls = float_of_int stats.nb_reads /. nb_entriesf in
+    let write_amplification_size =
+      float_of_int stats.bytes_written /. (entry_sizef *. nb_entriesf)
+    in
+    let write_amplification_calls =
+      float_of_int stats.nb_writes /. nb_entriesf
+    in
+    let ops_per_sec = nb_entriesf /. time in
+    let mbs_per_sec = entry_sizef *. nb_entriesf /. 1_048_576. /. time in
+    {
+      time;
+      ops_per_sec;
+      mbs_per_sec;
+      read_amplification_calls;
+      read_amplification_size;
+      write_amplification_calls;
+      write_amplification_size;
+    }
+
+  let pp_result fmt result =
+    Format.fprintf fmt
+      "Total time: %f@\n\
+       Operations per second: %f@\n\
+       Mbytes per second: %f@\n\
+       Read amplification in syscalls: %f@\n\
+       Read amplification in bytes: %f@\n\
+       Write amplification in syscalls: %f@\n\
+       Write amplification in bytes: %f" result.time result.ops_per_sec
+      result.mbs_per_sec result.read_amplification_calls
+      result.read_amplification_size result.write_amplification_calls
+      result.write_amplification_size
+end
+
+let make_bindings_pool nb_entries =
+  Array.init nb_entries (fun _ ->
+      let k = Context.Key.v () in
+      let v = Context.Value.v () in
+      (k, v))
+
+let bindings_pool = ref [||]
+
+let absent_bindings = ref [||]
+
+module Index = struct
+  module Index = Index_unix.Private.Make (Context.Key) (Context.Value)
+
+  let add_metrics =
+    let no_tags x = x in
+    fun () -> Metrics.add src no_tags (fun m -> m (Stats.get ()))
+
+  let write ~with_metrics ?(with_flush = false) bindings rw =
+    Array.iter
+      (fun (k, v) ->
+        Index.replace rw k v;
+        if with_flush then Index.flush rw;
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read ~with_metrics bindings r =
+    Array.iter
+      (fun (k, _) ->
+        ignore (Index.find r k : Context.Value.t);
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read_absent ~with_metrics bindings r =
+    Array.iter
+      (fun (k, _) ->
+        try ignore (Index.find r k : Context.Value.t)
+        with Not_found -> if with_metrics then add_metrics ())
+      bindings
+
+  let write_random ~with_metrics t () = write ~with_metrics !bindings_pool t
+
+  let write_seq ~with_metrics t =
+    let bindings = Array.copy !bindings_pool in
```

Comment:
I'm not sure about this. The point of `write_seq` is to see if an order on the keys has a better performance than `write_random`. But in  `write_seq` there are two (potentially big) arrays in the memory instead of one as in `write_random` and it affects the performance. 

Maybe sort the bindings_pool and have a flag which tells if you need to regenerate the array or not, in case `write_random` is called after a sort? Or order the benchmarks before calling them, so that the sorting benchmarks are called last?

Reply:
I solved this by having a third array in memory, dedicated to sorted benchs